src/defrag.c

```diff
@@ -844,299 +887,474 @@ int defragLaterItem(dictEntry *de, unsigned long *cursor, long long endtime, int
-             * check if we reached the time limit.
-             * But regardless, don't start a new db in this loop, this is because after
-             * the last db we call defragOtherGlobals, which must be done in one cycle */
-            if (all_stages_finished || ++iterations > 16 || server.stat_active_defrag_hits - prev_defragged > 512 ||
-                server.stat_active_defrag_scanned - prev_scanned > 64) {
-                /* Quit if all stages were finished or timeout. */
-                if (all_stages_finished || ustime() > endtime) {
-                    quit = 1;
-                    break;
-                }
-                iterations = 0;
-                prev_defragged = server.stat_active_defrag_hits;
-                prev_scanned = server.stat_active_defrag_scanned;
-            }
-        } while (!all_stages_finished && !quit);
-    } while (!quit);
+    if (!server.active_defrag_enabled) {
+        // Defrag has been disabled while running
+        endDefragCycle();
+        return AE_NOMORE;
+    }
+
+    if (hasActiveChildProcess()) {
+        // If there's a child process, pause the defrag, polling until the child completes.
+        defrag.timeproc_end_time = 0; // prevent starvation recovery
+        return 100;
+    }
+
+    monotime starttime = getMonotonicUs();
+    monotime endtime = starttime + computeDefragCycleUs();
+
+    mstime_t latency;
+    latencyStartMonitor(latency);
+
+    if (!defrag.current_stage) {
+        defrag.current_stage = listNodeValue(listFirst(defrag.remaining_stages));
+        listDelNode(defrag.remaining_stages, listFirst(defrag.remaining_stages));
+        // Initialize the stage with endtime==0
+        doneStatus status = defrag.current_stage->stage_fn(0, defrag.current_stage->target, defrag.current_stage->privdata);
+        serverAssert(status == DEFRAG_NOT_DONE); // Initialization should always return DEFRAG_NOT_DONE
+    }
+
+    doneStatus status = defrag.current_stage->stage_fn(endtime, defrag.current_stage->target, defrag.current_stage->privdata);
+    if (status == DEFRAG_DONE) {
+        zfree(defrag.current_stage);
+        defrag.current_stage = NULL;
+    }
 
     latencyEndMonitor(latency);
     latencyAddSampleIfNeeded("active-defrag-cycle", latency);
 
-update_metrics:
-    if (server.active_defrag_running > 0) {
-        if (server.stat_last_active_defrag_time == 0) elapsedStart(&server.stat_last_active_defrag_time);
-    } else if (server.stat_last_active_defrag_time != 0) {
-        server.stat_total_active_defrag_time += elapsedUs(server.stat_last_active_defrag_time);
-        server.stat_last_active_defrag_time = 0;
+    if (defrag.current_stage || listLength(defrag.remaining_stages) > 0) {
+        return computeDelayMs(endtime);
+    } else {
+        /* The timeproc will go away when we return AE_NOMORE.
+         * Mark this now, indicating a normal termination. */
+        defrag.timeproc_id = AE_DELETED_EVENT_ID;
```

Comment:
You're right, it was a bit weird.  I was clearing `timeproc_id` so that `endDefragCycle` would recognize defrag as not-running and would consider it a normal termination.  I fixed this by adding an explicit parameter for `normal_termination`, and then clear `timeproc_id` in one place.